bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
+                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                Err(VMError::from((bad_instruction, error)))
+            }
+        }
+    }
+}
+
+impl CellKind for u8 {
+    fn wrapping_increment(&mut self) {
+        *self = self.wrapping_add(1);
+    }
+
+    fn wrapping_decrement(&mut self) {
+        *self = self.wrapping_sub(1);
+    }
+
+    fn set_value(&mut self, value: u8) {
+        *self = value;
+    }
+
+    fn get_value(&self) -> u8 {
+        *self
+    }
+}
+
+/// Error types that the [VirtualMachine] can emit
+#[derive(Debug, PartialEq, Eq)]
+pub enum VMError {
+    /// The head ran off the start of the tape
+    HeadUnderrun(LocalisedInstruction),
+    /// The head ran off the end of the (non-auto-extending) tape
+    HeadOverrun(LocalisedInstruction),
+    /// Reading a byte from stdio went bloop
+    ReadError(LocalisedInstruction, String),
+    /// Writing a byte from stdio went bloop
+    WriteError(LocalisedInstruction, String),
+}
+
+impl std::error::Error for VMError {}
+
+impl Display for VMError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::HeadOverrun(program_instruction) => {
+                write!(
+                    f,
+                    "Head Overrun Error occured at line {} column {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num()
+                )
+            }
+            Self::HeadUnderrun(program_instruction) => {
+                write!(
+                    f,
+                    "Head Underrun Error occured at line {} column {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num()
+                )
+            }
+            Self::ReadError(program_instruction, error) => {
+                write!(
+                    f,
+                    "Read Error occured at line {} column {}: {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num(),
+                    error
+                )
+            }
+            Self::WriteError(program_instruction, error) => {
+                write!(
+                    f,
+                    "Write Error occured at line {} column {}: {}",
+                    program_instruction.line_num(),
+                    program_instruction.column_num(),
+                    error
+                )
+            }
+        }
+    }
+}
+
+impl From<(LocalisedInstruction, std::io::Error)> for VMError {
+    fn from(value: (LocalisedInstruction, std::io::Error)) -> Self {
+        let bad_instruction = value.0;
+        let error_msg = value.1.to_string();
+        if bad_instruction.instruction() == Instruction::Input {
+            VMError::ReadError(bad_instruction, error_msg)
+        } else {
+            VMError::WriteError(bad_instruction, error_msg)
+        }
+    }
+}
```

Comment:
You *could* replace all of this with use of the `thiserror` library to a greater or lesser extent.  The `From` impl might be a tad harder to get done with that library though at least you'd cover the `Display` without too much issue.  It'd be neater too.